include/anvill/Arch.h

```diff
@@ -0,0 +1,233 @@
+    llvm::Type &type, std::vector<bool> &reserved,
+    const std::vector<RegisterConstraint> &register_constraints);
+
+  // For each element of the struct, try to allocate it to a register, if all of
+  // them can be allocated, then return that allocation. Otherwise return a
+  // nullptr.
+  std::unique_ptr<std::vector<anvill::ValueDecl>> TryReturnThroughRegisters(
+      const llvm::StructType &st,
+      const std::vector<RegisterConstraint> &constraints);
+
+  const remill::Arch* arch;
+
+ private:
+  llvm::CallingConv::ID identity;
+};
+
+class X86_64_SysV : public CallingConvention {
+ public:
+  X86_64_SysV(const remill::Arch *arch) : CallingConvention(llvm::CallingConv::X86_64_SysV, arch) {}
+  virtual ~X86_64_SysV() = default;
+  void AllocateSignature(FunctionDecl &fdecl, const llvm::Function &func);
+  std::vector<ParameterDecl> BindParameters(const llvm::Function &function);
+  std::vector<ValueDecl> BindReturnValues(const llvm::Function &function);
+  void BindReturnStackPointer(FunctionDecl &fdecl, const llvm::Function &func);
+
+ private:
+  const std::vector<RegisterConstraint> parameter_register_constraints = {
+      RegisterConstraint({
+          VariantConstraint("DIL", kTypeIntegral, kMaxBit8),
+          VariantConstraint("DI", kTypeIntegral, kMaxBit16),
+          VariantConstraint("EDI", kTypeIntegral, kMaxBit32),
+          VariantConstraint("RDI", kTypeIntegral, kMaxBit64),
+      }),
+      RegisterConstraint({
+          VariantConstraint("SIL", kTypeIntegral, kMaxBit8),
+          VariantConstraint("SI", kTypeIntegral, kMaxBit16),
+          VariantConstraint("ESI", kTypeIntegral, kMaxBit32),
+          VariantConstraint("RSI", kTypeIntegral, kMaxBit64),
+      }),
+      RegisterConstraint({
+          VariantConstraint("DL", kTypeIntegral, kMaxBit8),
+          VariantConstraint("DX", kTypeIntegral, kMaxBit16),
+          VariantConstraint("EDX", kTypeIntegral, kMaxBit32),
+          VariantConstraint("RDX", kTypeIntegral, kMaxBit64),
+      }),
+      RegisterConstraint({
+          VariantConstraint("CL", kTypeIntegral, kMaxBit8),
+          VariantConstraint("CX", kTypeIntegral, kMaxBit16),
+          VariantConstraint("ECX", kTypeIntegral, kMaxBit32),
+          VariantConstraint("RCX", kTypeIntegral, kMaxBit64),
+      }),
+      RegisterConstraint({
+          VariantConstraint("R8L", kTypeIntegral, kMaxBit8),
+          VariantConstraint("R8W", kTypeIntegral, kMaxBit16),
+          VariantConstraint("R8D", kTypeIntegral, kMaxBit32),
+          VariantConstraint("R8", kTypeIntegral, kMaxBit64),
+      }),
+      RegisterConstraint({
+          VariantConstraint("R9L", kTypeIntegral, kMaxBit8),
+          VariantConstraint("R9W", kTypeIntegral, kMaxBit16),
+          VariantConstraint("R9D", kTypeIntegral, kMaxBit32),
+          VariantConstraint("R9", kTypeIntegral, kMaxBit64),
+      }),
+
+      RegisterConstraint({VariantConstraint("XMM0", kTypeFloat, kMaxBit128)}),
```

Comment:
Vectors of floats/integers can be passed over xmm regs in 64 bit. https://gcc.godbolt.org/z/BfgENx

Reply:
Fixed 75dfa9a